### PR TITLE
Drone Based GCS

### PR DESCRIPTION
**Pull request by ntlhui**
Working drone-based GCS support

### CODE REVIEW

File: RctGcs/rctGCS.py [hunk 1]
```diff
@@ -106,6 +105,7 @@ def configSetup() -> Path:
 def main():
     start_timestamp = dt.datetime.now()
     log_file = start_timestamp.strftime('%Y.%m.%d.%H.%M.%S_gcs.log')
+    log_file = '0gcs.log'
```
Comment (hgrehm) [inline RctGcs/rctGCS.py:108]: timestamp and filename based on it are unnecessary

File: RctGcs/config.py [hunk 2]
```diff
@@ -214,9 +189,7 @@ def qgis_prefix_set(self, value: Any) -> None:
 
     @classmethod
     def __get_qgis_path(cls) -> Path:
-        if "_CONDA_ROOT" not in os.environ:
-            raise RuntimeError("Not a conda environment")
-        pkgs_dir = Path(os.environ['_CONDA_ROOT']).joinpath('pkgs')
+        pkgs_dir = Path.home().joinpath('miniconda3', 'pkgs')
```
Comment (hgrehm) [inline RctGcs/config.py:192]: Explicitly disallows anaconda as opposed to miniconda; disregard if this is desired

File: RctGcs/ui/display.py [hunk 3]
```diff
@@ -1,47 +1,50 @@
...
-import RctGcs.rctCore
-from RctGcs.config import get_config_path, get_instance
-from RctGcs.ui.controls import *
-from RctGcs.ui.map import *
+from RctGcs.config import ConnectionMode, get_config_path, get_instance
+from RctGcs.rctCore import (EXTS_STATES, OUTPUT_DIR_STATES, RCT_STATES,
+                            SDR_INIT_STATES, Events, MAVModel, NoActiveModel)
+from RctGcs.ui.controls import CollapseFrame, SystemSettingsControl
+from RctGcs.ui.map import MapOptions, StaticMap, WebMap
+from RctGcs.ui.popups import ConnectionDialog, UserPopups
```
Comment (hgrehm) [inline RctGcs/ui/display.py:16]: ConfigDialog missing (gets called on line 532 when user clicks "Edit Configuration Settings", last dropdown item in UI)

File: RctGcs/droneSimulator.py [hunk 4]
```diff
@@ -1051,11 +1002,7 @@ def __connection_handler(self, connection, id):
 def main():
     parser = argparse.ArgumentParser(
         description='Radio Collar Tracker Payload Control Simulator')
-    parser.add_argument('--port', type=int, default=9000)
-    parser.add_argument('--protocol', type=str,
-                        choices=['udp', 'tcp'], required=True)
-    parser.add_argument('--target', type=str, default='127.0.0.1',
-                        help='Target IP Address.  Use 255.255.255.255 for broadcast, 127.0.0.1 for local')
+    parser.add_argument('--spec', type=str, required=True)
     parser.add_argument('--clients', type=int, default=1)
```
Comment (hgrehm) [inline RctGcs/droneSimulator.py:1006]: new usage not reflected in README

File: RctGcs/rctGCS.py (hunk 1 above)
Comment (ntlhui) [reply RctGcs/rctGCS.py]: Resolved in 073c7e70b161e8ed94be64dd9ec874f96ad41076

File: RctGcs/config.py (hunk 2 above)
Comment (ntlhui) [reply RctGcs/config.py:192]: Fixed in 073c7e70b161e8ed94be64dd9ec874f96ad41076

**Comment by hgrehm** [comment]
Having difficulties testing this with the simulator. Here's the traceback from my most recent attempt, which failed as I tried to connect the GCS to the sim:
```
2023-56-28 14:07:51.541: ERROR:rctComms.mavComms: Exception during callback
Traceback (most recent call last):
  File "C:\Users\hgreh\miniconda3\envs\rctGCS\lib\site-packages\RCTComms\comms.py", line 1222, in execute_cb
    cb_(**kwargs)
  File "C:\Users\hgreh\workspace\radio_collar_tracker_gcs\RctGcs\droneSimulator.py", line 562, in __do_get_options
    packet = RCTComms.comms.rctOptionsPacket(scope, **self.PP_options)
TypeError: keywords must be strings
{<Options.DSP_PING_WIDTH: 'DSP_ping_width'>: 27.0, <Options.DSP_PING_SNR: 'DSP_ping_snr'>: 0.1, <Options.DSP_PING_MAX: }
```

Curious if you're seeing this too? I can troubleshoot on my end if it might be an issue with versions or something else local to my machine. I don't see why an Enum shouldn't be allowed as a key since it's immutable...

File: RctGcs/ui/display.py (hunk 3 above)
Comment (ntlhui) [reply RctGcs/ui/display.py]: fixed in f6d403757f76ff3a41d209b67bffa41204e4be63

File: RctGcs/droneSimulator.py (hunk 4 above)
Comment (ntlhui) [reply RctGcs/droneSimulator.py:1006]: fixed in 5b5bffc8e2b96fb41c03d3cdfdc81f50e7a32920

**Comment by ntlhui** [comment]
> 

Fixed in 8eb890764935f7a09d0f8d16c8d5530c8bc435fc

All options are now of type `Dict[Options, Any]`.  You cannot keyword expand into `kwargs` if the keys are not `str`.  All option packets now accept a full dict instead of kwargs.

File: RctGcs/ui/popups.py
Comment (hgrehm) [inline RctGcs/ui/popups.py:1]: ConfigDialog and ConfigDialogPage still use `connection_port` and `connection_addr`

File: RctGcs/ui/controls.py [hunk 5]
```diff
@@ -353,17 +246,18 @@ def submit_gui_option_vars(self, scope: int):
...
         try:
-            cntr_freq = int(self.option_vars['SDR_center_freq'].text())
-            samp_freq = int(self.option_vars['SDR_sampling_freq'].text())
-            sdr_gain = float(self.option_vars['SDR_gain'].text())
+            cntr_freq = int(self.option_vars[Options.SDR_CENTER_FREQ].text())
+            samp_freq = int(self.option_vars[Options.SDR_SAMPLING_FREQ].text())
+            sdr_gain = float(self.option_vars[Options.SDR_GAIN].text())
         except ValueError:
             UserPopups.show_warning("Please enter center and sampling frequences and SDR gain settings.")
             return
```
Comment (hgrehm) [inline RctGcs/ui/controls.py:263]: unused method

File: RctGcs/ui/controls.py [hunk 6]
```diff
@@ -262,16 +177,20 @@ def __advanced_settings(self):
...
-        if abs(var - cntr_freq) > samp_freq:
+        try:
+            model = MAVModel.get_model()
+        except NoActiveModel:
+            return
+        cntr_freq = model.getOption(Options.SDR_CENTER_FREQ)
+        samp_freq = model.getOption(Options.SDR_SAMPLING_FREQ)
+        if abs(value - cntr_freq) > samp_freq:
             return False
         return True
```
Comment (hgrehm) [inline RctGcs/ui/controls.py:195]: unused method--should these checks be performed upon `add_entry` (in `widgets.py`)?

File: RctGcs/droneSimulator.py [hunk 7]
```diff
@@ -566,10 +544,10 @@ def __do_set_frequency(self, packet: RCTComms.comms.rctSETFCommand, addr: str):
 
         # Nyquist check
         for freq in frequencies:
-            if abs(freq - self.PP_options['SDR_center_freq']) > self.PP_options['SDR_sampling_freq']:
+            if abs(freq - self.PP_options[Options.SDR_CENTER_FREQ]) > self.PP_options[Options.SDR_SAMPLING_FREQ]:
```
Comment (hgrehm) [inline RctGcs/droneSimulator.py:547]: related to [previous comment](https://github.com/UCSD-E4E/radio_collar_tracker_gcs/pull/76/files#r1278143389)--Ideally we stop the user from sending these to the sim at all. Raising an error after the user's already connected seems like an extreme way to handle bad user input. Not that having a check here as a last resort is a bad idea though.

**Pull request closed without merging**